### PR TITLE
Configure with meta/config.

### PR DESCRIPTION
**Pull request by icemac**
Add support for Python 3.9.

See https://github.com/zopefoundation/meta/pull/94 for the needed changes in meta/config.

### CODE REVIEW

**Review by jamadden — COMMENTED** [review]
The tests are crashing because coverage can't find a plugin:
```
Run coverage run -p -m unittest discover -s src
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.6.13/x64/bin/coverage", line 8, in <module>
    sys.exit(main())
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/coverage/cmdline.py", line 871, in main
    status = CoverageScript().command_line(argv)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/coverage/cmdline.py", line 588, in command_line
    return self.do_run(options, args)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/coverage/cmdline.py", line 743, in do_run
    self.coverage.start()
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/coverage/control.py", line 532, in start
    self._init()
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/coverage/control.py", line 275, in _init
    self._plugins = Plugins.load_plugins(self.config.plugins, self.config, self._debug)
  File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/coverage/plugin_support.py", line 41, in load_plugins
    __import__(module)
ModuleNotFoundError: No module named 'coverage_python_version'
```

File: CHANGES.rst [hunk 1]
```diff
@@ -5,7 +5,7 @@
 5.4.1 (unreleased)
 ==================
 
-- Nothing changed yet.
+- Add support for Python 3.9.
```
[the remark this answers is not in the document]
Comment (icemac) [reply CHANGES.rst]: I changed the change log entry and hopefully fixed the tests in https://github.com/zopefoundation/zope.interface/pull/243/commits/183bd2d73984c12927b1bbf106c8a6221939c172.

**Comment by icemac** [comment]
Now CI fails at pushing data to coveralls.io because `coverage-python-version` is needed there too. (See https://github.com/zopefoundation/zope.interface/pull/243/checks?check_run_id=2515733925) 

This step is done via the GitHub action https://github.com/AndreMiras/coveralls-python-action and I do not see a chance there to install `coverage-python-version` beforehand.

It seems that we either have to remove `coverage-python-version` or find another way to publish to coveralls.io. @jamadden Do you know why this GitHub action is used instead of installing the Python package and publishing via calling its script?

**Comment by mgedmin** [comment]
> It seems that we either have to remove coverage-python-version or find another way to publish to coveralls.io.

I wonder that nobody else complained about this problem in https://github.com/AndreMiras/coveralls-python-action/issues.  Do people rarely use coverage plugins?

For my personal projects I've been uploading to coveralls.io by using the usual `pip install coveralls && coveralls` in the build script, e.g. https://github.com/mgedmin/check-python-versions/blob/57aa8665d74fee20a56c179608a1c0d79d8b13c0/.github/workflows/build.yml#L60-L65

One issue with that approach is that you need a recent enough version of https://pypi.org/p/coveralls to support GITHUB_TOKEN, and that version no longer supports Python 2.

**Comment by jamadden** [comment]
> One issue with that approach is that you need a recent enough version of https://pypi.org/p/coveralls to support GITHUB_TOKEN, and that version no longer supports Python 2.

That's one of the major reasons I've been using the docker version everywhere. 

> Do people rarely use coverage plugins?

I think that's right too.

**Comment by mgedmin** [comment]
I've filed an issue: https://github.com/AndreMiras/coveralls-python-action/issues/17

**Comment by icemac** [comment]
Let's see if there will be a reaction to AndreMiras/coveralls-python-action#17. Otherwise I'd drop the usage of `coverage-python-version` as it is not used here in `zope.interface`.

**Comment by mgedmin** [comment]
Oh, we're not actually using the coverage-python-version plugin here?  Then let's by all means drop it.

**Comment by icemac** [comment]
Dropped `coverage-python-version`, let's see if GHA gets green now.

File: .coveragerc [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+    .tox/pypy*/site-packages/
 
 [report]
 show_missing = true
+precision = 2
 exclude_lines =
+    except ImportError:
+    if __name__ == '__main__':
     pragma: no cover
-    class I[A-Z]\w+\((Interface|I[A-Z].*)\):
```
[the remark this answers is not in the document]
Comment (icemac) [reply .coveragerc:11]: I hopefully fixed all the places where this line ignored methods by changing `pass` to a docstring.

There is still https://coveralls.io/builds/39516428/source?filename=src%2Fzope%2Finterface%2Fcommon%2Fcollections.py#L221 which I do not understand as I thought we have coverage for all Python versions. (There are some other cases which look the same.)

File: .meta.toml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    "    coverage run -p -m sphinx -b doctest -d docs/_build/doctrees docs docs/_build/doctest",
+    ]
+coverage-command = "coverage combine"
+coverage-additional = [
+    "depends = py27,py27-pure,py35,py35-pure,py36,py36-pure,py37,py37-pure,py38,py38-pure,py39,py39-pure,pypy,pypy3,docs",
+    "parallel_show_output = true",
+    ]
+
+[coverage]
+fail-under = 98
```
[the remark this answers is not in the document]
Comment (icemac) [reply .meta.toml:47]: I doubt we actually had 100 % there are two files without any coverage:

* https://coveralls.io/builds/39516428/source?filename=src%2Fzope%2Finterface%2Fcommon%2Ftests%2Fbasemapping.py
* https://coveralls.io/builds/39516428/source?filename=src%2Fzope%2Finterface%2F_flatten.py

[the remark this answers is not in the document]
Comment (icemac) [reply .meta.toml:47]: Currently we have 99.25 % coverage, so I am going to update it here to 99.

File: .meta.toml [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    "    unzip -n dist/zope.interface-*whl -d src",
+    "    pip install -U -e .[test]",
+    "- name: Run tests and report coverage",
+    "  # Once with C extensions, once without. Yes, this runs them",
+    "  # twice on PyPy.",
+    "  run: |",
+    "    coverage run -p -m unittest discover -s src",
+    "    PURE_PYTHON=1 coverage run -p -m unittest discover -s src",
+    "    coverage combine",
+    "    coverage report -i",
```
[the remark this answers is not in the document]
Comment (icemac) [reply .meta.toml:92]: I always have to look up the short names for `coverage report` options, so I am going to use the long ones here.

File: tox.ini [hunk 5]
```diff
@@ -1,67 +1,77 @@
...
+# Until repoze.sphinx.autointerface supports Sphinx 4.x we cannot use it:
+deps =
+    Sphinx < 4
+    coverage
+extras = docs
+commands =
+    sphinx-build -b html -d docs/_build/doctrees docs docs/_build/html
+    coverage run -p -m sphinx -b doctest -d docs/_build/doctrees docs docs/_build/doctest
+
+[testenv:pure]
```
[the remark this answers is not in the document]
Comment (icemac) [reply tox.ini:40]: Thank you for looking into this, I am going to find a way to fix it.

File: tox.ini [hunk 6]
```diff
@@ -1,67 +1,77 @@
+# Generated from:
+# https://github.com/zopefoundation/meta/tree/master/config/c-code
 [tox]
+minversion = 3.18
 envlist =
-    py27,py27-pure,
-    py35,py35-pure,
-    py36,
-    py37,
-    py38,py38-cext,
```
[the remark this answers is not in the document]
Comment (icemac) [reply tox.ini:7]: `py38` now behaves the same way as `py38-cext` did previously: It sets `PURE_PYTHON` to `0`.

[the remark this answers is not in the document]
Comment (mgedmin) [reply tox.ini:7]: I didn't notice!

I worry that it's a bit implicit.  In time py38 will be EOL and if nobody notices, we'll lose the PURE_PYTHON=0 env.

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -248,7 +248,7 @@ jobs:
           coverage run -p -m unittest discover -s src
           PURE_PYTHON=1 coverage run -p -m unittest discover -s src
           coverage combine
-          coverage report -i
+          coverage report --ignore-errors --show-missing
```
Comment (mgedmin) [inline .github/workflows/tests.yml:251]: :100:  (I had to look up `-i` before)

Actually I was so busy looking up what -i meant that I forgot to ask why we needed it here?

File: tox.ini (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgedmin) [reply tox.ini:40]: In the main [testenv], this should work:

```dosini
setenv =
    pure: PURE_PYTHON=1
    !pure: PURE_PYTHON=0
    ZOPE_INTERFACE_STRICT_IRO=1
```

although if we want pypy to also default to PURE_PYTHON=1, then, hmm

```dosini
setenv =
    pure,pypy: PURE_PYTHON=1
    !pure-!pypy: PURE_PYTHON=0
    ZOPE_INTERFACE_STRICT_IRO=1
```

if I understand the [docs](https://tox.readthedocs.io/en/latest/config.html#complex-factor-conditions) correctly.

File: tox.ini (hunk 6 above)
[the remark this answers is not in the document]
Comment (mgedmin) [reply tox.ini:7]: Never mind, *all* non-pure toxenvs pass PURE_PYTHON=0, they're not going extinct by accident.

File: .github/workflows/tests.yml (hunk 7 above)
Comment (icemac) [reply .github/workflows/tests.yml:251]: I thought previously there was a comment in the file to explain it. But I was not able to find it. Maybe in another repository.

**Comment by icemac** [comment]
@mgedmin You requested changes to this PR. I think I fixed them. Are you okay with them?

**Review by mgedmin — APPROVED** [review]
lgtm

**Comment by icemac** [comment]
Thank you for reviewing this PR. 😃

**Pull request merged**